### PR TITLE
Support target artifact selection in JSON I/O

### LINKED ISSUE #2897: Support output selection in standard json

### PR DESCRIPTION
**Pull request by axic**
Fixes #2897. Depends on #2729.

### CODE REVIEW

File: libsolidity/interface/StandardCompiler.cpp [hunk 1]
```diff
@@ -146,6 +146,47 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
 	return output;
 }
 
+bool isTargetRequired(Json::Value const& _targets, string const& _target)
+{
+	for (auto const& target: _targets)
+		if (target == _target)
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp]: Need to support sub matches, e.g. `evm.assembly` should be enabled by both `evm.assembly`, `evm`, `evm.*`, etc.

**Comment by chriseth** [comment]
What is left to do here? Can you work on this next, @axic?

**Comment by axic** [comment]
Reviewed it again, I think the API (`isTargetRequired`) and the part using it is clear enough, but the actual implementation is broken, hence the tests failures.

It also doesn't support wildcards properly and shortest match.

**Comment by axic** [comment]
Need to add tests for matching different combinations (`evm`, `evm.*`, `evm.assembly`, etc.)

File: test/libsolidity/StandardCompiler.cpp [hunk 2]
```diff
@@ -181,6 +181,12 @@ BOOST_AUTO_TEST_CASE(basic_compilation)
 			"fileA": {
 				"content": "contract A { }"
 			}
+		},
+		"outputSelection": {
+			"fileA": {
+				"A": [ "abi", "devdoc", "userdoc", "evm", "metadata" ]
```
Comment (axic) [inline test/libsolidity/StandardCompiler.cpp]: This is annoying, one of the reasons for the failing tests was a missing comma here at the end 😕

File: libsolidity/interface/StandardCompiler.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/interface/StandardCompiler.cpp]: @chriseth do you have any code suggestion for doing so? E.g. requiring `evm.assembly` matches both the user supplied `evm.assembly`, `evm.*` and `evm`, without hardcoding these paths.

File: Changelog.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 Features:
  * Parser: Better error message for unexpected trailing comma in parameter lists.
- * Standard JSON: Support the ``outputSelection`` field for selective compilation of supplied sources.
+ * Standard JSON: Support the ``outputSelection`` field for selective compilation.
```
Comment (chriseth) [inline Changelog.md]: Please move to 0.4.19

**Comment by axic** [comment]
Damn, the reason for failure was a typo, `file` vs `_file`. This just shows the bad naming :)

File: libsolidity/interface/StandardCompiler.cpp [hunk 4]
```diff
@@ -131,6 +131,62 @@ StringMap createSourceList(Json::Value const& _input)
...
+/// @_contract is the current contract
+/// @_target is the current target name
+///
+/// Returns true if the _targets has a match for the requested target in the specific file / contract.
+///
+/// In @_targets the use of '*' as a wildcard is permitted.
+///
+/// @TODO optimise this. Perhaps flatten the structure upfront.
+///
+bool isTargetRequired(Json::Value const& _targets, string const& _file, string const& _contract, string const& _target)
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp]: Better naming: `isArtifactRequired(_outputSelection, _file, _contract, _artifactName)`

**Comment by axic** [comment]
Depends on https://github.com/ethereum/solc-js/pull/159.

**Comment by axic** [comment]
@chriseth This is working now (as confirmed by circleci). It still needs support for shorcuts (e.g. `evm.bytecode` matching `evm.bytecode.object`)

**Comment by axic** [comment]
@chriseth can you review this? it may be worth even merging it in this state, though more tests cannot hurt.

File: solc/jsonCompiler.cpp [hunk 5]
```diff
@@ -128,6 +128,11 @@ string compile(StringMap const& _sources, bool _optimize, CStyleReadFileCallback
 	input["settings"]["optimizer"]["enabled"] = _optimize;
 	input["settings"]["optimizer"]["runs"] = 200;
 
+	// Enable all SourceUnit-level outputs.
+	input["settings"]["outputSelection"]["*"][""][0] = "*";
```
Comment (chriseth) [inline solc/jsonCompiler.cpp:132]: Does `[0]` automatically make it an array?

File: libsolidity/interface/StandardCompiler.cpp [hunk 6]
```diff
@@ -131,6 +131,70 @@ StringMap createSourceList(Json::Value const& _input)
 	return sources;
 }
 
+bool isTargetRequired(Json::Value const& _targets, string const& _target)
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: Perhaps better `requested` than `required`?

File: libsolidity/interface/StandardCompiler.cpp [hunk 7]
```diff
@@ -131,6 +131,70 @@ StringMap createSourceList(Json::Value const& _input)
...
+		return false;
+
+	for (auto const& file: { _file, string("*") })
+		if (_targets.isMember(file) && _targets[file].isObject())
+		{
+			if (_contract.empty())
+			{
+				/// Special case for SourceUnit-level targets (such as AST)
+				if (
+					_targets[file].isMember("") &&
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: Can you extract this condition into a function and also use it in line 129?

File: solc/jsonCompiler.cpp (hunk 5 above)
Comment (axic) [reply solc/jsonCompiler.cpp:132]: Yep. Had an ugly large code before to do it manually.

File: libsolidity/interface/StandardCompiler.cpp (hunk 7 above)
Comment (axic) [reply libsolidity/interface/StandardCompiler.cpp]: 129?

File: libsolidity/interface/StandardCompiler.cpp (hunk 6 above)
Comment (axic) [reply libsolidity/interface/StandardCompiler.cpp]: Sure, but also please see my other naming suggestion below.

File: libsolidity/interface/StandardCompiler.cpp (hunk 7 above)
Comment (chriseth) [reply libsolidity/interface/StandardCompiler.cpp]: Sorry, 179

File: libsolidity/interface/StandardCompiler.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/interface/StandardCompiler.cpp]: Fine with that, but would perhaps use `isArtifactRequested`

**Comment by axic** [comment]
@chriseth I think this should be ready now?

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
@chriseth added two more commits to simplify/optimize it even more (last 3 commits should be squashed). Please check those two.

**Comment by chriseth** [comment]
Looks good!

**Pull request merged**